File: 142.linked-list-cycle-ii.java
/*
 * @lc app=leetcode id=142 lang=java
 *
 * [142] Linked List Cycle II
 */

// @lc code=start
/**
 * Definition for singly-linked list. class ListNode { int val; ListNode next;
 * ListNode(int x) { val = x; next = null; } }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        if (head == null)
            return null;
        // 1. is there a cycle?
        boolean flag = false;
        ListNode runner = head, walker = head, end = head;
        while (runner.next != null && runner.next.next != null) {
            runner = runner.next.next;
            walker = walker.next;
            if (runner == walker) {
                flag = true;
                end = runner;
                // System.out.println(end.val);
                break;
            }
        }
        if (flag == false)
            return null;

        // 2. break cycle at end node
        // consider it as find the intersection of two lists

        // 2.1 get the lengths
        ListNode head1 = head, head2 = end.next;
        int len1 = 0, len2 = 0;
        while (head1 != end) {
            len1++;
            head1 = head1.next;
        }
        while (head2 != end) {
            len2++;
            head2 = head2.next;
        }

        // 2.2 set the start node
        head1 = head;
        head2 = end.next;
        while (len1 > len2) {
            len1--;
            head1 = head1.next;
        }
        while (len2 > len1) {
            len2--;
            head2 = head2.next;
        }

        // 2.3 find the intersection
        while (head1 != head2) {
            head1 = head1.next;
            head2 = head2.next;
        }
        return head1;
    }
}
// @lc code=end

File: 142.linked-list-cycle-ii.py
#
# @lc app=leetcode id=142 lang=python3
#
# [142] Linked List Cycle II
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        slow = fast = head

        # use slow and fast pointer to find some point in the cycle, p
        while fast and fast.next:
            slow, fast = slow.next, fast.next.next
            if slow == fast:
                break
        else:
            return None  # if not (fast and fast.next): return None
        p = slow

        # each time head and p move 1 step, until they meet at the start of cycle.
        while head != p:
            head, p = head.next, p.next
        return head


# @lc code=end
